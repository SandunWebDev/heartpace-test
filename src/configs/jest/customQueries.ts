import {
	queryHelpers,
	buildQueries,
	Matcher,
	MatcherOptions,
} from '@testing-library/react';

// Custom query to directly match Cypress "data-cy" elements.
const queryAllByDataCy = (
	container: HTMLElement,
	id: Matcher,
	options?: MatcherOptions,
) => queryHelpers.queryAllByAttribute('data-testid', container, id, options);

const getMultipleError = (_c: Element | null, dataCyValue: string) =>
	`Found multiple elements with the data-test attribute of: ${dataCyValue}`;
const getMissingError = (_c: Element | null, dataCyValue: string) =>
	`Unable to find an element with the data-cy attribute of: ${dataCyValue}`;

const [
	queryByDataCy,
	getAllByDataCy,
	getByDataCy,
	findAllByDataCy,
	findByDataCy,
] = buildQueries(queryAllByDataCy, getMultipleError, getMissingError);

export {
	queryByDataCy,
	queryAllByDataCy,
	getByDataCy,
	getAllByDataCy,
	findAllByDataCy,
	findByDataCy,
};
