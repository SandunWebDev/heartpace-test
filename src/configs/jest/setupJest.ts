import '@testing-library/jest-dom';
import ResizeObserver from 'resize-observer-polyfill';

// Mocking ResizeObserver for "Recharts"
global.ResizeObserver = ResizeObserver;
