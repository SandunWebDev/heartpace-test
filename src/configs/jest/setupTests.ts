import '@testing-library/jest-dom';

// Polyfilling ResizeObserver. Currently mainly needed for "Recharts"
import ResizeObserver from 'resize-observer-polyfill';
global.ResizeObserver = ResizeObserver;
