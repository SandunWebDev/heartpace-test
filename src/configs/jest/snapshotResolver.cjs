// Making "snapshot"  files are create besides the test file.

const resolveSnapshotPath = (
	testPath, // Path of the test file being test3ed
	snapshotExtension, // The extension for snapshots (.snap usually)
) => {
	const snapshotFilePath = testPath + snapshotExtension; // Ex. some.test.tsx + '.snap'
	return snapshotFilePath;
};

const resolveTestPath = (
	snapshotFilePath, // The filename of the snapshot (i.e. some.test.tsx.snap)
	snapshotExtension, // The extension for snapshots (.snap)
) => {
	const testPath = snapshotFilePath
		.replace(snapshotExtension, '')
		.replace('__snapshots__/', ''); //Remove the .snap
	return testPath;
};

/* Used to validate resolveTestPath(resolveSnapshotPath( {this} )) */
const testPathForConsistencyCheck = 'some.test.tsx';

module.exports = {
	resolveSnapshotPath,
	resolveTestPath,
	testPathForConsistencyCheck,
};
