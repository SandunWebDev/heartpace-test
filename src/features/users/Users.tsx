import { useEffect } from 'react';
import Box from '@mui/material/Box';
import Typography from '@mui/material/Typography';
import PeopleIcon from '@mui/icons-material/People';

import {
	useAppDispatch,
	useAppSelector,
	shallowEqual,
} from '../../services/redux/hooks';
import {
	usersActions,
	usersSelectors,
} from '../../services/redux/slices/users/usersSlice';
import LoadingView from '../../components/loaders/LoadingView';
import UsersTable from './UsersTable';

export default function Users() {
	const dispatch = useAppDispatch();
	const { userList, getAllUsersReqStatus, getAllUsersReqError } =
		useAppSelector(usersSelectors.selectGetAllUsersReqState, shallowEqual);

	useEffect(() => {
		dispatch(usersActions.getAllUsers());
	}, [dispatch]);

	const isDataLoading = getAllUsersReqStatus === 'LOADING';
	const isDataLoadingError = getAllUsersReqStatus === 'ERROR';

	return (
		<Box>
			<Box sx={{ marginBottom: '30px' }}>
				<Typography
					variant='h3'
					color='textPrimary'
					sx={{ display: 'flex', alignItems: 'center', gap: '5px' }}>
					<PeopleIcon fontSize='large' sx={{ fontSize: 'inherit' }} />
					Users
				</Typography>
				<Typography variant='subtitle2' color='textsecondary'>
					All Company Users
				</Typography>
			</Box>

			<LoadingView
				isLoading={isDataLoading}
				isError={Boolean(isDataLoadingError)}
				loadingText='Loading User List'
				errorText={getAllUsersReqError}
			/>

			{!isDataLoading && !isDataLoadingError && (
				<UsersTable
					userList={userList}
					getAllUsersReqStatus={getAllUsersReqStatus}
					getAllUsersReqError={getAllUsersReqError}
				/>
			)}
		</Box>
	);
}
