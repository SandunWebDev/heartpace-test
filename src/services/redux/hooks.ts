// This file serves as a central hub for re-exporting pre-typed Redux hooks.
// Use these throughout the app instead of plain `useDispatch` and `useSelector`.
import { useDispatch, useSelector } from 'react-redux';
import type { AppDispatch, RootState } from './store';

export const useAppDispatch = useDispatch.withTypes<AppDispatch>();
export const useAppSelector = useSelector.withTypes<RootState>();

// Use "shallowEqual" with "useAppSelector" when selectors returnning object. See https://stackoverflow.com/a/78727535
export { shallowEqual } from 'react-redux';
