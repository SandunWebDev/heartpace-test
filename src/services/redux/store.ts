import type { Action, ThunkAction } from '@reduxjs/toolkit';
import { configureStore, combineReducers } from '@reduxjs/toolkit';
import usersReducer, {
	initialState as usersSliceInitialState,
} from './slices/users/usersSlice';

export const initialPreloadedState = {
	users: usersSliceInitialState,
};

const rootReducer = combineReducers({
	users: usersReducer,
});

export const setupStore = (preloadedState?: Partial<RootState>) => {
	return configureStore({
		reducer: rootReducer,
		preloadedState,
	});
};

export type RootState = ReturnType<typeof rootReducer>;
export type AppStore = ReturnType<typeof setupStore>;
export type AppDispatch = AppStore['dispatch'];
export type AppThunk<ThunkReturnType = void> = ThunkAction<
	ThunkReturnType,
	RootState,
	unknown,
	Action
>;
