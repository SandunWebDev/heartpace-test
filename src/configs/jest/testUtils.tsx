import React, { ReactElement } from 'react';
import { render, queries, within, RenderOptions } from '@testing-library/react';
import { Provider as ReduxProvider } from 'react-redux';
import { ToastContainer } from 'react-toastify';

import MuiThemeProvider from '../theme/MuiThemeProvider';
import { store } from '../../services/redux/store';
import * as customQueries from './customQueries';

// All the default providers should goes here. So we don't have to wrap these for each test.
const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
	return (
		<MuiThemeProvider>
			<ReduxProvider store={store}>
				<ToastContainer />
				{children}
			</ReduxProvider>
		</MuiThemeProvider>
	);
};

const allQueries = {
	...queries,
	...customQueries,
};

// Re-exporting "React Testing Library's" original "render, screen, within" with our Custom Wrappers and Custom Queries.
const customRender = (
	ui: ReactElement,
	options?: Omit<RenderOptions, 'wrapper' | 'queries'>,
) => render(ui, { wrapper: AllTheProviders, queries: allQueries, ...options });
const customScreen = within(document.body, allQueries);
const customWithin = (element: HTMLElement) => within(element, allQueries);

export * from '@testing-library/react';
export {
	customRender as render,
	customScreen as screen,
	customWithin as within,
};
