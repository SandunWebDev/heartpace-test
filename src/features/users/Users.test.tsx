import { Provider as ReduxProvider } from 'react-redux';
import { store } from '../../services/redux/store';

import { render, screen, cleanup } from '../../configs/jest/testUtils';
import Users from './Users';

describe('<Users/>', () => {
	beforeEach(() => {
		render(
			<ReduxProvider store={store}>
				<Users />
			</ReduxProvider>,
		);
	});

	afterEach(cleanup);

	it('Users snapshot', () => {
		const { asFragment } = render(
			<ReduxProvider store={store}>
				<Users />
			</ReduxProvider>,
		);
		expect(asFragment()).toMatchSnapshot();
	});

	it('Users get rendered', () => {
		const Users = screen.getByText('Users');
		expect(Users).toBeInTheDocument();
	});
});
