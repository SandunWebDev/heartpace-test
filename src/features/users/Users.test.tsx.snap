// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Users/> Users snapshot 1`] = `
<DocumentFragment>
  <div
    class="Toastify"
  />
  <div
    class="MuiBox-root css-0"
  >
    <div
      class="MuiBox-root css-1xn3edc"
    >
      <h3
        class="MuiTypography-root MuiTypography-h3 css-158hdt-MuiTypography-root"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge css-1qpx589-MuiSvgIcon-root"
          data-testid="PeopleIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16 11c1.66 0 2.99-1.34 2.99-3S17.66 5 16 5s-3 1.34-3 3 1.34 3 3 3m-8 0c1.66 0 2.99-1.34 2.99-3S9.66 5 8 5 5 6.34 5 8s1.34 3 3 3m0 2c-2.33 0-7 1.17-7 3.5V19h14v-2.5c0-2.33-4.67-3.5-7-3.5m8 0c-.29 0-.62.02-.97.05 1.16.84 1.97 1.97 1.97 3.45V19h6v-2.5c0-2.33-4.67-3.5-7-3.5"
          />
        </svg>
        Users
      </h3>
      <h6
        class="MuiTypography-root MuiTypography-subtitle2 css-qxsh4m-MuiTypography-root"
      >
        All Company Users
      </h6>
    </div>
    <div
      class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded css-i9zgug-MuiPaper-root"
    >
      <div
        class="MuiBox-root css-1ff8r59"
      >
        <div
          class="MuiBox-root css-0"
        >
          <span
            class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-1wxecgq-MuiCircularProgress-root"
            role="progressbar"
            style="width: 40px; height: 40px;"
          >
            <svg
              class="MuiCircularProgress-svg css-54pwck-MuiCircularProgress-svg"
              viewBox="22 22 44 44"
            >
              <circle
                class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-960umw-MuiCircularProgress-circle"
                cx="44"
                cy="44"
                fill="none"
                r="20.2"
                stroke-width="3.6"
              />
            </svg>
          </span>
          <h6
            class="MuiTypography-root MuiTypography-subtitle1 css-i5usgo-MuiTypography-root"
          >
            Loading User List
          </h6>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
